### PR TITLE
Updated OS type identification

### PR DESCRIPTION
**Pull request by interkosmos**
As noted in the source code and discussed in #144, the OS detection of `get_os_type()` in `fpm/src/environment.f90` is not accurate. Please review this request, which checks environment variables and the existance of some files in order to determine the OS more precisely.

### CODE REVIEW

**Comment by LKedward** [comment]
Thanks @interkosmos, this looks like an improvement. Do note that you'll need to update any functions that use `get_os_type` such as [`list_files` in `fpm.f90`](https://github.com/fortran-lang/fpm/blob/cd10478e26c428e599a1cecc99b03b6b0c1292d3/fpm/src/fpm.f90#L32) to consider all different OS behaviours.

I'm hoping #155 will be merged soon, and this further increase the usage of `get_os_type` (mainly for filesystem/command_line interaction) so it may be best to wait for this to be merged?

I have no experience with Solaris or BSD so I don't know what quirks need to be specifically addressed, but we may run into the problem that we cannot test them in the CI.

**Comment by awvwgk** [comment]
BSD could be tested using Cirrus CI, which provides [FreeBSD images](https://cirrus-ci.org/guide/FreeBSD/). Grouping it together with OSX and Linux will probably be fine for most use cases right now, to avoid another CI provider for this project.

Testing cygwin is not (easily) possible on the [GH actions windows image](https://github.com/actions/virtual-environments/blob/main/images/win/Windows2019-Readme.md), which means another CI provider would be required for this purpose (like [Appveyor](https://www.appveyor.com/docs/windows-images-software/)).

No idea were to find a CI solution for Solaris.

**Comment by interkosmos** [comment]
In most cases, the behaviour of `OS_LINUX`, `OS_MACOS`, `OS_CYGWIN`, `OS_SOLARIS`, and `OS_FREEBSD` should be identical, if we stick to POSIX standards. I can’t say if there are any CI providers that feature OpenSolaris-based images (OpenIndiana, illumos, Tribblix, …).

File: fpm/src/environment.f90
Comment (certik) [inline fpm/src/environment.f90:23]:
```suggestion
        !! names. If this fails too, check the existence of files that can be
```

**Review by certik — APPROVED** [review]
This looks good. I'll let @LKedward merge #155 first, and then this.

**Review by milancurcic — APPROVED** [review]
Thank you!

**Comment by LKedward** [comment]
@interkosmos, would you mind also updating the following uses of `get_os_type()` in order to prevent the `select case` statements from silently falling through?

https://github.com/fortran-lang/fpm/blob/e02171d28bb783bb419b44ef453ca56286b389a6/fpm/src/fpm_filesystem.f90#L51-L56

https://github.com/fortran-lang/fpm/blob/e02171d28bb783bb419b44ef453ca56286b389a6/fpm/src/fpm_filesystem.f90#L117-L124

https://github.com/fortran-lang/fpm/blob/e02171d28bb783bb419b44ef453ca56286b389a6/fpm/src/fpm_filesystem.f90#L148-L158

https://github.com/fortran-lang/fpm/blob/e02171d28bb783bb419b44ef453ca56286b389a6/fpm/src/fpm_command_line.f90#L66-L73

**Comment by certik** [comment]
The current failure is (on Windows):
```
Command line: gfortran -c -Jbuild\gfortran_debug\fpm -Ibuild\gfortran_debug\toml-f -Wall -Wextra -Wimplicit-interface -fPIC -fmax-errors=1 -g -fbounds-check -fcheck-array-temporaries -fbacktrace -o build/gfortran_debug/fpm/fpm_filesystem.o src\fpm_filesystem.f90
Exit code: 1
Stderr:
src\fpm_filesystem.f90:2:10:

     use :: fpm_environment, only: get_os_type, &
          1
Fatal Error: Can't open module file 'fpm_environment.mod' for reading at (1): No such file or directory
compilation terminated.
```
So that needs to be fixed before we can merge it.

**Comment by interkosmos** [comment]
@certik: does this happend because `fpm_filesystem.f90` is compiled before `fpm_environment.f90` by fpm?

**Comment by LKedward** [comment]
@interkosmos can you try reverting to the non colon form of the `use` statement - this may be a bug in Haskell fpm.

File: fpm/src/fpm_command_line.f90
Comment (LKedward) [inline fpm/src/fpm_command_line.f90:2]: Probably best to be consistent with all `use` statements now. (no colons)

**Review by LKedward — APPROVED** [review]
Great work @interkosmos. CI is now passing so this looks good to go! Only left one minor comment regarding a remaining colon use statement.

File: fpm/src/fpm_command_line.f90
Comment (LKedward) [reply fpm/src/fpm_command_line.f90:2]: (this is most likely affected by the same bug but compiles in the right order by happenchance)

**Review by certik — APPROVED** [review]
Looks good.

Next time I recommend to do changes like `" -> '` (and white space formatting) in a separate pull request, as those are unrelated to the "OS type identification" improvement. It makes it easier to review of what actually changed.

**Comment by interkosmos** [comment]
Please excuse the inconvenience.

**Pull request merged**
